numbat/src/typechecker/map_stack.rs

```diff
@@ -0,0 +1,80 @@
+use std::hash::Hash;
+use std::{borrow::Borrow, collections::HashMap};
+
+/// A stack of hash maps. All insertions affect the top hash map, preserving
+/// any entries in maps below. The `save` function can be used to push a new
+/// map on the top of the stack, in effect saving the current state of the map,
+/// which one can then restore with `restore`.
+#[derive(Debug, Clone)]
+pub(crate) struct MapStack<K, V> {
+    top: HashMap<K, V>,
```

Comment:
I wonder why you chose to keep this as a separate element in the struct instead of keeping it in the `Vec`? Doing so might simplify the implementation of some functions below, but probably has other disadvantages?

Reply:
Initially it was a bit more convenient to have it this way, because it saved some `unwrap/except` in a bunch of places. However now that the data structure is in its own file, it's not that different. I've changed the implementation to just use a stack with the function `restore` in charge of ensuring its not empty and `insert` panicking if it can't find anything at the top of the stack (which should be impossible because `restore` will panic before). 

Also that allowed me to see a bug in my implementation: the stack vector should be traversed in reverse order, I've fixed `iter_dict` and `iter_dict_mut`.